flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironment.java

```diff
@@ -0,0 +1,87 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.taskexecutor.partition;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.runtime.deployment.ResultPartitionDeploymentDescriptor;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.io.network.api.writer.ResultPartitionWriter;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.shuffle.ShuffleEnvironment;
+
+import java.util.Collection;
+import java.util.function.Consumer;
+
+/**
+ * Job-aware variation of the {@link ShuffleEnvironment} interface.
+ *
+ * @see ShuffleEnvironment
+ */
+public interface JobAwareShuffleEnvironment<P extends ResultPartitionWriter, G extends InputGate> extends ShuffleEnvironment<P, G> {
+
+	/**
+	 * Factory method for the task's {@link ResultPartitionWriter}s.
+	 *
+	 * @param jobId the job id, used for tracking which job the created partitions belongs to
+	 * @param taskName the task name, used for logs
+	 * @param executionAttemptID execution attempt id of the task
+	 * @param resultPartitionDeploymentDescriptors descriptors of the partition, produced by the task
+	 * @param outputGroup shuffle specific group for output metrics
+	 * @param buffersGroup shuffle specific group for buffer metrics
+	 * @return collection of the task's {@link ResultPartitionWriter}s
+	 */
+	Collection<P> createResultPartitionWriters(
+		JobID jobId,
+		String taskName,
+		ExecutionAttemptID executionAttemptID,
+		Collection<ResultPartitionDeploymentDescriptor> resultPartitionDeploymentDescriptors,
+		MetricGroup outputGroup,
+		MetricGroup buffersGroup);
+
+	/**
+	 * Releases local resources occupied with the given partitions.
+	 */
+	void releaseFinishedPartitions(JobID jobId, Collection<ResultPartitionID> partitionIds);
+
+	/**
+	 * Releases all local resources occupied with partitions belonging to the given job.
+	 */
+	void releaseAllFinishedPartitionsAndMarkJobInactive(JobID jobId);
+
+	/**
+	 * Returns whether there are any unreleased partitions for a given job.
+	 *
+	 * @return whether there are any unreleased partitions for the given job
+	 */
+	boolean hasPartitionsOccupyingLocalResources(JobID jobId);
+
+	/**
+	 * Marks the given job as active. If a partition is finished while the job is no longer active it will
+	 * be immediately released.
+	 */
+	void markJobActive(JobID jobId);
+
+	/**
+	 * Sets a listener that is called if a partition was failed or finished, and subsequently no more unreleased
+	 * partitions exist for a given job.
+	 */
+	void setPartitionFailedOrFinishedListener(Consumer<JobID> listener);
```

Comment:
Might be more explicit to define a separate interface with a SAM. It might not be super clear what this consumer actually does when looking at the call `listener.accept(jobId)`.